File: js/parser/exchange.js
import ccxt from "ccxt"

export class Exchange {
	constructor(id, marketType, name, region, cache) {
		this.id = id
		this.name = null
		this.region = region
		this.properties = null
		this.type = marketType
		this.stale = false

		if (marketType === "crypto") {
			this.properties = cache ? cache : new ccxt[id]()
			if (!cache) {
				try { this.properties.loadMarkets() } catch (e) { this.stale = false }
			}
			// USDⓈ-M
			if (id == "binanceusdm") this.name = "Binance Futures"
			// COIN-M
			else if (id == "binancecoinm") this.name = "Binance Futures COIN-M"
			else this.name = this.properties.name
		} else {
			this.properties = new StocksExchange(id)
			this.name = !name ? id.charAt(0).toUpperCase() + id.substring(1) : name
		}
	}

	toDict() {
		return {
			id: this.id,
			name: this.name,
			region: this.region,
			type: this.type,
		}
	}

	static fromDict(d, cache) {
		if (!d || Object.keys(d).length == 0) return null
		return new Exchange(d.id, d.type, d.name, d.region, cache)
	}
}

class StocksExchange {
	constructor(id) {
		this.id = id
		this.symbols = []
		this.markets = {}
		this.timeframes = ["1m"]
	}

	milliseconds() {
		return Date.now()
	}
}

File: py/parser/TickerParser/exchange.py
from time import time
from traceback import format_exc
import ccxt


class Exchange(object):
	def __init__(self, id, marketType, name=None, region=None):
		self.id = id
		self.name = None
		self.region = region
		self.properties = None
		self.type = marketType

		if self.type == "crypto":
			self.properties = getattr(ccxt, id)()
			if id == "binanceusdm": self.name = "Binance Futures" # USDⓈ-M
			elif id == "binancecoinm": self.name = "Binance Futures COIN-M"
			else: self.name = self.properties.name
		else:
			self.properties = StocksExchange(id)
			self.name = id.title() if name is None else name

	def to_dict(self):
		return {
			"id": self.id,
			"name": self.name,
			"region": self.region,
			"type": self.type
		}

	@staticmethod
	def from_dict(d):
		if d is None or not d: return None
		return Exchange(d.get("id"), d.get("type"), d.get("name"), d.get("region"))

	def __hash__(self):
		return hash(self.id)

	def __str__(self):
		return f"{self.name} [id: {self.id}]"

class StocksExchange(object):
	def __init__(self, id):
		self.id = id
		self.symbols = []
		self.markets = {}
		self.timeframes = ["1m"]

	def milliseconds(self):
		return int(time() * 1000)